File: packages/go/whatsapp-go/webhooks.go
package whatsapp

import (
	"crypto/hmac"
	"crypto/sha256"
	"crypto/subtle"
	"encoding/hex"
	"encoding/json"
	"fmt"
)

type InboundMessage struct {
	From         string
	ID           string
	Timestamp    string
	Type         string
	Text         string
	CallbackData string
	Raw          map[string]any
}

func VerifyWebhookChallenge(query map[string]string, verifyToken string) (string, error) {
	if verifyToken == "" {
		return "", &ConfigurationError{Message: "WHATSAPP_VERIFY_TOKEN e obrigatorio"}
	}
	mode := first(query, "hub.mode", "hub_mode")
	token := first(query, "hub.verify_token", "hub_verify_token")
	challenge := first(query, "hub.challenge", "hub_challenge")
	if mode == "subscribe" && token == verifyToken {
		return challenge, nil
	}
	return "", &ValidationError{APIError: &APIError{Status: 403, Message: "Token de verificacao de webhook invalido"}}
}

func MetaSignatureHeader(rawBody, appSecret string) string {
	mac := hmac.New(sha256.New, []byte(appSecret))
	_, _ = mac.Write([]byte(rawBody))
	return "sha256=" + hex.EncodeToString(mac.Sum(nil))
}

func VerifyMetaSignature(rawBody, signatureHeader, appSecret string) (bool, error) {
	if appSecret == "" {
		return false, &ConfigurationError{Message: "WHATSAPP_APP_SECRET e obrigatorio"}
	}
	if signatureHeader == "" {
		return false, nil
	}
	expected := MetaSignatureHeader(rawBody, appSecret)
	if len(expected) != len(signatureHeader) {
		return false, nil
	}
	if subtle.ConstantTimeCompare([]byte(expected), []byte(signatureHeader)) == 1 {
		return true, nil
	}
	return false, nil
}

func VerifyMetaSignatureOrThrow(rawBody, signatureHeader, appSecret string) error {
	ok, err := VerifyMetaSignature(rawBody, signatureHeader, appSecret)
	if err != nil {
		return err
	}
	if !ok {
		return &SignatureError{Message: "Assinatura X-Hub-Signature-256 invalida"}
	}
	return nil
}

func ParseInboundMessages(payload any) []InboundMessage {
	var out []InboundMessage
	root := asRecord(payload)
	entries, _ := root["entry"].([]any)
	for _, entry := range entries {
		er := asRecord(entry)
		changes, _ := er["changes"].([]any)
		for _, change := range changes {
			cr := asRecord(change)
			value := asRecord(cr["value"])
			messages, _ := value["messages"].([]any)
			for _, msg := range messages {
				rec := asRecord(msg)
				text := ""
				if t, ok := rec["text"].(map[string]any); ok {
					text, _ = t["body"].(string)
				}
				cb, _ := rec["biz_opaque_callback_data"].(string)
				out = append(out, InboundMessage{
					From:         fmt.Sprint(zero(rec["from"])),
					ID:           fmt.Sprint(zero(rec["id"])),
					Timestamp:    fmt.Sprint(zero(rec["timestamp"])),
					Type:         fmt.Sprint(zero(rec["type"])),
					Text:         text,
					CallbackData: cb,
					Raw:          rec,
				})
			}
		}
	}
	return out
}

func first(m map[string]string, keys ...string) string {
	for _, k := range keys {
		if v, ok := m[k]; ok {
			return v
		}
	}
	return ""
}

func zero(v any) any {
	if v == nil {
		return ""
	}
	return v
}

// Ensure json is referenced if payload arrives as []byte.
func parseJSON(raw []byte) any {
	var payload any
	if err := json.Unmarshal(raw, &payload); err != nil {
		return map[string]any{"raw": string(raw)}
	}
	return payload
}

File: packages/ruby/whatsapp/lib/ordex_whatsapp/webhooks.rb
# frozen_string_literal: true

require "openssl"

module OrdexWhatsApp
  module Webhooks
    module_function

    def verify_challenge(query, verify_token)
      raise ConfigurationError, "WHATSAPP_VERIFY_TOKEN e obrigatorio" if verify_token.to_s.strip.empty?

      query = stringify(query)
      mode = query["hub.mode"] || query["hub_mode"]
      token = query["hub.verify_token"] || query["hub_verify_token"]
      challenge = query["hub.challenge"] || query["hub_challenge"] || ""
      return challenge.to_s if mode == "subscribe" && token == verify_token

      raise ValidationError.new("Token de verificacao de webhook invalido", status: 403)
    end

    def signature_header(raw_body, app_secret)
      hex = OpenSSL::HMAC.hexdigest("SHA256", app_secret, raw_body)
      "sha256=#{hex}"
    end

    def verify_signature(raw_body, signature_header, app_secret)
      raise ConfigurationError, "WHATSAPP_APP_SECRET e obrigatorio" if app_secret.to_s.strip.empty?
      return false if signature_header.nil? || signature_header.empty?

      expected = signature_header(raw_body, app_secret)
      return false unless expected.bytesize == signature_header.bytesize

      OpenSSL.fixed_length_secure_compare(expected, signature_header)
    end

    def verify_signature!(raw_body, signature_header, app_secret)
      return true if verify_signature(raw_body, signature_header, app_secret)

      raise SignatureError, "Assinatura X-Hub-Signature-256 invalida"
    end

    def parse_inbound_messages(payload)
      out = []
      return out unless payload.is_a?(Hash)

      Array(payload["entry"] || payload[:entry]).each do |entry|
        next unless entry.is_a?(Hash)

        Array(entry["changes"] || entry[:changes]).each do |change|
          next unless change.is_a?(Hash)

          value = change["value"] || change[:value] || {}
          Array(value["messages"] || value[:messages]).each do |msg|
            next unless msg.is_a?(Hash)

            text = msg["text"] || msg[:text]
            out << {
              from: (msg["from"] || msg[:from]).to_s,
              id: (msg["id"] || msg[:id]).to_s,
              timestamp: (msg["timestamp"] || msg[:timestamp]).to_s,
              type: (msg["type"] || msg[:type]).to_s,
              text: text.is_a?(Hash) ? (text["body"] || text[:body]) : nil,
              callback_data: msg["biz_opaque_callback_data"] || msg[:biz_opaque_callback_data],
              raw: msg
            }
          end
        end
      end
      out
    end

    def stringify(hash)
      return {} if hash.nil?

      hash.each_with_object({}) { |(k, v), acc| acc[k.to_s] = v }
    end
  end
end
